src/pkg/services/m365/api/channels_pager.go

```diff
@@ -1,43 +1,223 @@
 package api
 
 import (
+	"context"
+
 	"github.com/microsoftgraph/msgraph-sdk-go/models"
+	"github.com/microsoftgraph/msgraph-sdk-go/teams"
+
+	"github.com/alcionai/corso/src/internal/common/ptr"
+	"github.com/alcionai/corso/src/internal/m365/graph"
+	"github.com/alcionai/corso/src/pkg/logger"
 )
 
 // ---------------------------------------------------------------------------
-// item pager
+// channel message pager
 // ---------------------------------------------------------------------------
 
-type ChannelMessageDeltaEnumerator interface {
-	DeltaGetPager
-	ValuesInPageLinker[models.ChatMessageable]
-	SetNextLinker
+var _ DeltaPager[models.ChatMessageable] = &channelMessageDeltaPageCtrl{}
+
+type channelMessageDeltaPageCtrl struct {
+	resourceID, channelID string
+	gs                    graph.Servicer
+	builder               *teams.ItemChannelsItemMessagesDeltaRequestBuilder
+	options               *teams.ItemChannelsItemMessagesDeltaRequestBuilderGetRequestConfiguration
+}
+
+func (p *channelMessageDeltaPageCtrl) SetNext(nextLink string) {
+	p.builder = teams.NewItemChannelsItemMessagesDeltaRequestBuilder(nextLink, p.gs.Adapter())
+}
+
+func (p *channelMessageDeltaPageCtrl) GetPage(
+	ctx context.Context,
+) (DeltaPageLinker, error) {
+	resp, err := p.builder.Get(ctx, p.options)
+	return resp, graph.Stack(ctx, err).OrNil()
 }
 
-// TODO: implement
-// var _ ChannelMessageDeltaEnumerator = &messagePageCtrl{}
+func (p *channelMessageDeltaPageCtrl) Reset(context.Context) {
+	p.builder = p.gs.
+		Client().
+		Teams().
+		ByTeamId(p.resourceID).
+		Channels().
+		ByChannelId(p.channelID).
+		Messages().
+		Delta()
+}
+
+func (p *channelMessageDeltaPageCtrl) ValuesIn(l PageLinker) ([]models.ChatMessageable, error) {
+	return getValues[models.ChatMessageable](l)
+}
+
+func (c Channels) NewChannelMessageDeltaPager(
+	teamID, channelID string,
+) *channelMessageDeltaPageCtrl {
+	builder := c.Stable.
+		Client().
+		Teams().
+		ByTeamId(teamID).
+		Channels().
+		ByChannelId(channelID).
+		Messages().
+		Delta()
+
+	options := &teams.ItemChannelsItemMessagesDeltaRequestBuilderGetRequestConfiguration{
+		Headers: newPreferHeaders(preferPageSize(maxNonDeltaPageSize)),
+	}
+
+	return &channelMessageDeltaPageCtrl{
+		resourceID: teamID,
+		channelID:  channelID,
+		builder:    builder,
+		gs:         c.Stable,
+		options:    options,
+	}
+}
+
+// GetChannelMessagesDelta fetches a delta of all messages in the channel.
+func (c Channels) GetChannelMessagesDelta(
+	ctx context.Context,
+	teamID, channelID, prevDelta string,
+) ([]models.ChatMessageable, DeltaUpdate, error) {
+	var (
+		vs               = []models.ChatMessageable{}
+		pager            = c.NewChannelMessageDeltaPager(teamID, channelID)
+		invalidPrevDelta = len(prevDelta) == 0
+		newDeltaLink     string
+	)
+
+	// Loop through all pages returned by Graph API.
+	for {
+		page, err := pager.GetPage(graph.ConsumeNTokens(ctx, graph.SingleGetOrDeltaLC))
+		if graph.IsErrInvalidDelta(err) {
```

Comment:
hmmm I know this probably follows similar code we have, but looking at it there's the assumption that we get the invalid delta token error on the first call to the API. It's probably fine to just check for `IsErrInvalidDelta`, but we should probably clear `vs` when we see it

Reply:
Ah, good call, thank you.